internal/controller/replication.storage/volumegroupreplication_controller.go

```diff
@@ -33,21 +57,536 @@ type VolumeGroupReplicationReconciler struct {
+				if err != nil {
+					logger.Error(err, "failed to fetch pvc from VGR status")
+					return reconcile.Result{}, err
+				}
+
+				if err = RemoveOwnerFromPVCAnnotation(r.Client, logger, pvc, replicationv1alpha1.VolumeGroupReplicationNameAnnotation); err != nil {
+					logger.Error(err, "Failed to remove VolumeReplication annotation from PersistentVolumeClaim")
+
+					return reconcile.Result{}, err
+				}
+
+				if err = RemoveFinalizerFromPVC(r.Client, logger, pvc, vgrReplicationFinalizer); err != nil {
+					logger.Error(err, "Failed to remove VGR finalizer from PersistentVolumeClaim")
+					return reconcile.Result{}, err
+				}
+			}
+		}
+		if instance.Spec.VolumeReplicationName != "" {
+			req := types.NamespacedName{Name: instance.Spec.VolumeReplicationName, Namespace: req.Namespace}
+			err = r.Client.Get(ctx, req, vrObj)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					logger.Info("volumeReplication resource not found")
+				} else {
+					logger.Error(err, "failed to fetch dependent volumeReplication resource")
+					return reconcile.Result{}, err
+				}
+			} else {
+				err = r.Client.Delete(ctx, vrObj)
+				if err != nil {
+					logger.Error(err, "failed to delete dependent volumeReplication resource")
+					return reconcile.Result{}, err
+				}
+			}
+		}
+
+		if instance.Spec.VolumeGroupReplicationContentName != "" {
+			req := types.NamespacedName{Name: instance.Spec.VolumeGroupReplicationContentName}
+			err = r.Client.Get(ctx, req, vgrContentObj)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					logger.Info("volumeGroupReplicationContent resource not found")
+				} else {
+					logger.Error(err, "failed to fetch dependent volumeGroupReplicationContent resource")
+					return reconcile.Result{}, err
+				}
+			} else {
+				err = r.Client.Delete(ctx, vgrContentObj)
+				if err != nil {
+					logger.Error(err, "failed to delete dependent volumeGroupReplicationContent resource")
+					return reconcile.Result{}, err
+				}
+			}
+		}
+
+		// Just log error, and exit reconcile without error. The dependent resource will udpate the VGR
+		// to remove their names from the CR, that will trigger a reconcile.
+		if err = RemoveFinalizerFromVGR(r.Client, logger, instance); err != nil {
+			logger.Error(err, "failed to remove VolumeGroupReplication finalizer")
+		}
+
+		return reconcile.Result{}, nil
+	}
 
-	return ctrl.Result{}, nil
+	// Update VGR status based on VR Status
+	instance.Status.VolumeReplicationStatus = vrObj.Status
+	err = r.Client.Status().Update(ctx, instance)
+	if err != nil {
+		logger.Error(err, "failed to update volumeGroupReplication instance's status", "VGRName", instance.Name)
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *VolumeGroupReplicationReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	err := r.waitForGroupCrds()
```

Comment:
Please include a description of what all needs to be watched for in the comment of the function. If it is commented, there is at least some verification possible that everything was implemented.

Reply:
Added the comments at all the places, wherever applicable.